src/Stratis.Features.Collateral/ConsensusRules/VotingRequestFullValidationRule.cs

```diff
@@ -0,0 +1,85 @@
+        private readonly IFederationManager federationManager;
+
+        public VotingRequestFullValidationRule(IInitialBlockDownloadState ibdState, Network network, CounterChainNetworkWrapper counterChainNetwork, IFederationManager federationManager, VotingManager votingManager) : base()
+        {
+            this.ibdState = ibdState;
+            this.network = network;
+            this.counterChainNetwork = counterChainNetwork.CounterChainNetwork;
+            this.encoder = new JoinFederationRequestEncoder();
+            this.federationManager = federationManager;
+            this.votingManager = votingManager;
+        }
+
+        /// <summary>Checks that any voting requests that are present can be decoded and prohibits re-use of collateral addresses.</summary>
+        public override Task RunAsync(RuleContext context)
+        {
+            if (this.ibdState.IsInitialBlockDownload())
+            {
+                this.Logger.LogTrace("(-)[SKIPPED_IN_IBD]");
+                return Task.CompletedTask;
+            }
+
+            foreach (Transaction transaction in context.ValidationContext.BlockToValidate.Transactions)
+            {
+                CheckTransaction(transaction);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public void CheckTransaction(Transaction transaction)
+        {
+            if (transaction.IsCoinBase || transaction.IsCoinStake)
+                return;
+
+            // This will raise a consensus error if there is a voting request and it can't be decoded.
+            JoinFederationRequest request = JoinFederationRequestBuilder.Deconstruct(transaction, this.encoder);
+            if (request == null)
+                return;
+
+            if (this.federationManager.IsMultisigMember(request.PubKey))
+            {
+                this.Logger.LogTrace("(-)[INVALID_MULTISIG_VOTING]");
+                PoAConsensusErrors.VotingRequestInvalidMultisig.Throw();
+            }
+
+            // Check collateral amount.
+            var collateralAmount = CollateralFederationMember.GetCollateralAmountForPubKey((PoANetwork)this.network, request.PubKey);
+
+            if (request.CollateralAmount.ToDecimal(MoneyUnit.BTC) != collateralAmount)
+            {
+                this.Logger.LogTrace("(-)[INVALID_COLLATERAL_REQUIREMENT]");
+                PoAConsensusErrors.InvalidCollateralRequirement.Throw();
+            }
+
+            // Prohibit re-use of collateral addresses.
+            Script script = PayToPubkeyHashTemplate.Instance.GenerateScriptPubKey(request.CollateralMainchainAddress);
+            string collateralAddress = script.GetDestinationAddress(this.counterChainNetwork).ToString();
+            CollateralFederationMember owner = this.federationManager.CollateralAddressOwner(this.votingManager, VoteKey.AddFederationMember, collateralAddress);
+            if (owner != null && owner.PubKey != request.PubKey)
```

Comment:
I think that the `owner.PubKey != request.PubKey` can actually be dropped? As the rule is that we abolsutely cannot have a duplicate address regardless of pubkey?

Reply:
It's just checking that a given collateral address is never associated with more than one public key. Removing that test would simply prevent the request initiator from wasting money if such a valid relationship already exists, and as such I agree.